metadata/src/main/java/org/apache/kafka/metadata/util/SnapshotReason.java

```diff
@@ -16,19 +16,48 @@
  */
 package org.apache.kafka.metadata.util;
 
-public enum SnapshotReason {
-    UnknownReason("unknown reason"),
-    MaxBytesExceeded("max bytes were exceeded"),
-    MetadataVersionChanged("metadata version was changed");
+import java.util.stream.Collectors;
+import java.util.stream.StreamSupport;
+import org.apache.kafka.server.common.MetadataVersion;
 
-    private final String snapshotReason;
+/**
+ * Reason for generating a snapshot.
+ */
+public final class SnapshotReason {
+    static public final SnapshotReason UNKNOWN = new SnapshotReason("unknown reason");
+
+    static public SnapshotReason maxBytesExceeded(long bytes, long maxBytes) {
+        return new SnapshotReason(String.format("%s bytes exceeded the maximum bytes of %s", bytes, maxBytes));
+    }
+
+    static public SnapshotReason maxIntervalExceeded(long interval, long maxInterval) {
+        return new SnapshotReason(
+            String.format("%s ms exceeded the maximum snapshot interval of %s ms", interval, maxInterval)
+        );
+    }
+
+    static public SnapshotReason metadataVersionChanged(MetadataVersion metadataVersion) {
+        return new SnapshotReason(String.format("metadata version was changed to %s", metadataVersion));
+    }
+
+    /**
+     * Converts a collection of reasons into a string.
+     */
+    static public String stringFromReasons(Iterable<SnapshotReason> reasons) {
```

Comment:
I think you can use `Utils.join` as well.

Reply:
Done.